src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -0,0 +1,431 @@
+-- 'hasVertex' x empty       == False
+-- @
+empty :: AdjacencyMap a b
+empty = BAM Map.empty Map.empty
+
+-- | Constructs the bipartite graph comprising /a single isolated vertex/ in
+-- the left part.
+-- Complexity: /O(1)/ time and memory.
+--
+-- @
+-- 'leftAdjacencyMap' (leftVertex x)  == Map.'Map.singleton' x Set.'Set.empty'
+-- 'rightAdjacencyMap' (leftVertex x) == Map.'Map.empty'
+-- 'hasEdge' x y (leftVertex x)       == False
+-- 'hasLeftVertex' x (leftVertex x)   == True
+-- 'hasRightVertex' x (leftVertex x)  == False
+-- @
+leftVertex :: a -> AdjacencyMap a b
+leftVertex x = BAM (Map.singleton x Set.empty) Map.empty
+
+-- | Constructs the bipartite graph comprising /a single isolated vertex/ in
+-- the right part.
+-- Complexity: /O(1)/ time and memory.
+--
+-- @
+-- 'leftAdjacencyMap' (rightVertex x)  == Map.'Map.empty'
+-- 'rightAdjacencyMap' (rightVertex x) == Map.'Map.singleton' x Set.'Set.empty'
+-- 'hasEdge' x y (rightVertex y)       == False
+-- 'hasLeftVertex' x (rightVertex x)   == False
+-- 'hasRightVertex' x (rightVertex x)  == True
+-- @
+rightVertex :: b -> AdjacencyMap a b
+rightVertex y = BAM Map.empty (Map.singleton y Set.empty)
+
+-- | Constructs the bipartite graph comprising /a single isolated vertex/.
+-- Complexity: /O(1)/ time and memory.
+--
+-- @
+-- vertex (Left x)                == 'leftVertex' x
+-- vertex (Right x)               == 'rightVertex' x
+-- 'hasEdge' x y (vertex (Left x))  == False
+-- 'hasEdge' x y (vertex (Right y)) == False
+-- @
+vertex :: Either a b -> AdjacencyMap a b
+vertex (Left x)  = leftVertex x
+vertex (Right y) = rightVertex y
+
+-- | Constructs the bipartite graph comprising /a single edge/.
+-- Complexity: /O(1)/ time and memory.
+--
+-- @
+-- 'leftAdjacencyMap' (edge x y)  == Map.'Map.singleton' x (Set.'Set.singleton' y)
+-- 'rightAdjacencyMap' (edge x y) == Map.'Map.singleton' y (Set.'Set.singleton' x)
+-- 'hasEdge' x y (edge x y)       == True
+-- 'hasEdge' y x (edge x y)       == (x == y)
+-- @
+edge :: a -> b -> AdjacencyMap a b
+edge x y = BAM (Map.singleton x (Set.singleton y)) (Map.singleton y (Set.singleton x))
+
+-- | /Overlay/ two bipartite graphs. This is a commutative, associative and
+-- idempotent operation with the identity 'empty'.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- 'isEmpty'     (overlay x y) == 'isEmpty'   x   && 'isEmpty'   y
+-- 'hasVertex' z (overlay x y) == 'hasVertex' z x || 'hasVertex' z y
+-- 'vertexCount' (overlay x y) >= 'vertexCount' x
+-- 'vertexCount' (overlay x y) <= 'vertexCount' x + 'vertexCount' y
+-- 'edgeCount'   (overlay x y) >= 'edgeCount' x
+-- 'edgeCount'   (overlay x y) <= 'edgeCount' x   + 'edgeCount' y
+-- @
+overlay :: (Ord a, Ord b) => AdjacencyMap a b -> AdjacencyMap a b -> AdjacencyMap a b
+overlay (BAM lr1 rl1) (BAM lr2 rl2) = BAM (Map.unionWith Set.union lr1 lr2) (Map.unionWith Set.union rl1 rl2)
+
+-- | /Connect/ two bipartite graphs, not adding the inappropriate edges. This
```

Comment:
Make it clearer: "not adding the inappropriate edges" => "not adding edges between vertices in the same part".